packages/core/src/workflow/buffer-reducing.js

```diff
@@ -0,0 +1,126 @@
+import { Aggregate, Piece } from '@web3-storage/data-segment'
+
+import { decode as decodeBuffer, encodeBlock } from '../data/buffer.js'
+
+/**
+ * @typedef {import('@web3-storage/data-segment').PieceLink} PieceLink
+ * @typedef {import('../data/types.js').Buffer<PieceLink>} Buffer
+ * @typedef {import('../data/types.js').BufferedPiece<PieceLink>} BufferedPiece
+ * @typedef {import('../data/types.js').Aggregate<PieceLink>} Aggregate
+ */
+
+/**
+ * @param {object} props
+ * @param {import('@web3-storage/filecoin-api/types').Store<Buffer>} props.storeClient 
+ * @param {import('@web3-storage/filecoin-api/types').Queue<Buffer>} props.bufferQueueClient
+ * @param {import('@web3-storage/filecoin-api/types').Queue<Aggregate>} props.aggregateQueueClient
+ * @param {string[]} props.bufferRecords
+ * @param {number} props.maxAggregateSize
+ * @param {number} props.minAggregateSize
+ * @param {string} [props.groupId]
+ */
+export async function reduceBuffer ({
+  storeClient,
+  bufferQueueClient,
+  aggregateQueueClient,
+  bufferRecords,
+  maxAggregateSize,
+  minAggregateSize,
+  groupId
+}) {
+  const bufferReferences = await Promise.all(
+    bufferRecords.map((message) => getBuffer(message, storeClient))
+  )
+
+  // Check if one of the buffers failed to get
+  const bufferReferenceGetError = bufferReferences.find(get => get.error)
+  if (bufferReferenceGetError) {
+    return {
+      error: bufferReferenceGetError.error
+    }
+  }
+
+  // @ts-expect-error typescript does not understand with find that no error
+  const { storefront, group } = bufferReferences[0].ok
+
+  /** @type {BufferedPiece[]} */
+  const pieces = []
+  for (const b of bufferReferences) {
+    // eslint-disable-next-line unicorn/prefer-spread
+    pieces.concat(b.ok?.pieces || [])
+  }
```

Comment:
concat does not mutate any arrays it creates a new one so think this is not going to behave correctly.

Reply:
yes good catch, these were the tests I still had in progress. Adding final commit that fixes it